File: tools/audit_multiplayer_bridge.mjs
#!/usr/bin/env node
import fs from 'node:fs';
const p=process.argv[2];
if(!p){console.error('usage: audit_multiplayer_bridge.mjs server/index.js');process.exit(2)}
const t=fs.readFileSync(p,'utf8');
const ws=t.indexOf("wss.on('connection'");
const tcp=t.indexOf('net.createConnection',ws);
const handlerEnd=t.indexOf("const ping=setInterval",ws);
const checks={
  ws_connection_handler:ws>=0,
  tcp_created_inside_ws_handler:ws>=0&&tcp>ws&&(handlerEnd<0||tcp<handlerEnd),
  per_session_id:t.includes('const id=++sessionSeq'),
  session_map:t.includes('sessions.set(id'),
  ws_close_destroys_tcp:t.includes("ws.on('close'")&&t.includes('tcp.destroy()'),
  tcp_close_removes_session:t.includes("tcp.on('close'")&&t.includes('sessions.delete(id)')
};
const out={ok:Object.values(checks).every(Boolean),checks,
  meaning:'Structural audit only. Live multiplayer gameplay still requires concurrent player testing.'};
console.log(JSON.stringify(out,null,2));
process.exit(out.ok?0:2);

File: tools/audit_multiplayer_bridge.py
#!/usr/bin/env python3
from pathlib import Path
import argparse,json,re
ap=argparse.ArgumentParser();ap.add_argument("server_file");a=ap.parse_args()
t=Path(a.server_file).read_text(encoding="utf-8")
checks={
 "ws_connection_handler": "wss.on('connection'" in t,
 "tcp_created_inside_ws_handler": bool(re.search(r"wss\.on\('connection'.*?net\.createConnection",t,re.S)),
 "per_session_id": "const id=++sessionSeq" in t,
 "session_map": "sessions.set(id" in t,
 "ws_close_destroys_tcp": "ws.on('close'" in t and "tcp.destroy()" in t,
 "tcp_close_removes_session": "tcp.on('close'" in t and "sessions.delete(id)" in t
}
out={"ok":all(checks.values()),"checks":checks,
"meaning":"Structural audit only. Real multiplayer correctness still requires live concurrent-player testing."}
print(json.dumps(out,ensure_ascii=False,indent=2))
raise SystemExit(0 if out["ok"] else 2)
